### PR TITLE
Standardize list pattern lowering on `Index` constructor.

### LINKED ISSUE #57825: List-patterns: lowering assumes the conversion from int to Index will use the implicit conversion operator
In `BindLengthAndIndexerForListPattern` we do:
`_ = GetWellKnownTypeMember(WellKnownMember.System_Index__op_Implicit_FromInt32, diagnostics, syntax: node);`

From [Aleksey](https://github.com/dotnet/roslyn/pull/57318#pullrequestreview-808648391): Just to clarify. What bothers me here is the fact that lowering never uses this member explicitly, it relies on conversion classification instead, etc. So, we are making an assumption that all that complicated machinery ends up using the method. However, there is no actual guarantee that that would be the case. For example, if method is not marked with special name flag, it won't be considered as a valid conversion operator.

As part of this, we should confirm whether we want list-pattern to lower as `expr[0]` (using conversion from int to Index) or `expr[new Index(0)]` (using constructor).

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #57825.
Relevant spec adjustment https://github.com/dotnet/csharplang/pull/5484.

Relates to test plan: https://github.com/dotnet/roslyn/issues/51289

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@jcouv, @333fred, @dotnet/roslyn-compiler Please review.

**Review by 333fred — APPROVED** [review]
LGTM (commit 3)

**Comment by AlekseyTs** [comment]
@jcouv, @dotnet/roslyn-compiler For the second review.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs [hunk 1]
```diff
@@ -432,6 +432,24 @@ private enum PatternIndexOffsetLoweringStrategy
                 strategy = PatternIndexOffsetLoweringStrategy.UseAsIs;
                 return VisitExpression(operand);
             }
+            else if (unloweredExpr is BoundObjectCreationExpression { Constructor: MethodSymbol constructor, Arguments: { Length: 2 } arguments, ArgsToParamsOpt: { IsDefaultOrEmpty: true }, InitializerExpressionOpt: null } &&
+                     (object)constructor == _compilation.GetWellKnownTypeMember(WellKnownMember.System_Index__ctor) &&
+                     arguments[0] is { Type.SpecialType: SpecialType.System_Int32, ConstantValue.Value: int _ and >= 0 } index &&
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs:437]: >_ [](http://example.com/codeflow?start=108&length=1)

nit: could remove the discard designation

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs [hunk 2]
```diff
@@ -432,6 +432,24 @@ private enum PatternIndexOffsetLoweringStrategy
                 strategy = PatternIndexOffsetLoweringStrategy.UseAsIs;
                 return VisitExpression(operand);
             }
+            else if (unloweredExpr is BoundObjectCreationExpression { Constructor: MethodSymbol constructor, Arguments: { Length: 2 } arguments, ArgsToParamsOpt: { IsDefaultOrEmpty: true }, InitializerExpressionOpt: null } &&
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs:435]: >InitializerExpressionOpt: null [](http://example.com/codeflow?start=190&length=30)

nit: do we have a indexing test with an Index initialization with an initializer?

**Review by jcouv — APPROVED** [review]
LGTM Thanks! (iteration 3)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs:435]: > do we have a indexing test with an Index initialization with an initializer?

No, the type doesn't have any member that can be initialized, but I decided to have the check for completeness.

**Pull request merged**